File: app/src/main/java/net/ktnx/mobileledger/json/ParsedLedgerAccount.java
/*
 * Copyright © 2021 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.json;

import androidx.annotation.Nullable;

import net.ktnx.mobileledger.async.RetrieveTransactionsTask;
import net.ktnx.mobileledger.model.AmountStyle;
import net.ktnx.mobileledger.model.LedgerAccount;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public abstract class ParsedLedgerAccount {
    private String aname;
    private int anumpostings;
    public abstract List<SimpleBalance> getSimpleBalance();
    public String getAname() {
        return aname;
    }
    public void setAname(String aname) {
        this.aname = aname;
    }
    public int getAnumpostings() {
        return anumpostings;
    }
    public void setAnumpostings(int anumpostings) {
        this.anumpostings = anumpostings;
    }
    public LedgerAccount toLedgerAccount(RetrieveTransactionsTask task,
                                         HashMap<String, LedgerAccount> map) {
        task.addNumberOfPostings(getAnumpostings());
        final String accName = getAname();
        LedgerAccount acc = map.get(accName);
        if (acc != null)
            throw new RuntimeException(
                    String.format("Account '%s' already present", acc.getName()));
        String parentName = LedgerAccount.extractParentName(accName);
        ArrayList<LedgerAccount> createdParents = new ArrayList<>();
        LedgerAccount parent;
        if (parentName == null) {
            parent = null;
        }
        else {
            parent = task.ensureAccountExists(parentName, map, createdParents);
            parent.setHasSubAccounts(true);
        }
        acc = new LedgerAccount(accName, parent);
        map.put(accName, acc);

        String lastCurrency = null;
        float lastCurrencyAmount = 0;
        AmountStyle lastAmountStyle = null;
        for (SimpleBalance b : getSimpleBalance()) {
            task.throwIfCancelled();
            final String currency = b.getCommodity();
            final float amount = b.getAmount();
            final AmountStyle amountStyle = b.getAmountStyle();
            if (currency.equals(lastCurrency)) {
                lastCurrencyAmount += amount;
                // Keep the first amountStyle found for this currency
            }
            else {
                if (lastCurrency != null) {
                    acc.addAmount(lastCurrencyAmount, lastCurrency, lastAmountStyle);
                }
                lastCurrency = currency;
                lastCurrencyAmount = amount;
                lastAmountStyle = amountStyle;
            }
        }
        if (lastCurrency != null) {
            acc.addAmount(lastCurrencyAmount, lastCurrency, lastAmountStyle);
        }
        for (LedgerAccount p : createdParents)
            acc.propagateAmountsTo(p);

        return acc;
    }

    static public class SimpleBalance {
        private String commodity;
        private float amount;
        @Nullable
        private AmountStyle amountStyle;
        public SimpleBalance(String commodity, float amount) {
            this.commodity = commodity;
            this.amount = amount;
        }
        public SimpleBalance(String commodity, float amount, @Nullable AmountStyle amountStyle) {
            this.commodity = commodity;
            this.amount = amount;
            this.amountStyle = amountStyle;
        }
        public String getCommodity() {
            return commodity;
        }
        public void setCommodity(String commodity) {
            this.commodity = commodity;
        }
        public float getAmount() {
            return amount;
        }
        public void setAmount(float amount) {
            this.amount = amount;
        }
        @Nullable
        public AmountStyle getAmountStyle() {
            return amountStyle;
        }
        public void setAmountStyle(@Nullable AmountStyle amountStyle) {
            this.amountStyle = amountStyle;
        }
    }
}

File: app/src/main/kotlin/net/ktnx/mobileledger/json/ParsedLedgerAccount.kt
/*
 * Copyright © 2021 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.json

import net.ktnx.mobileledger.domain.model.Account
import net.ktnx.mobileledger.domain.model.AccountAmount
import net.ktnx.mobileledger.domain.model.AmountStyle

abstract class ParsedLedgerAccount {
    open var aname: String = ""
    open var anumpostings: Int = 0

    abstract fun getSimpleBalance(): List<SimpleBalance>

    /**
     * Convert to domain model Account.
     *
     * Note: This does NOT create parent accounts. Parent account creation
     * should be handled by the caller (e.g., TransactionSyncerImpl).
     */
    open fun toDomain(): Account {
        val level = aname.count { it == ':' }
        val balances = getSimpleBalance()
        return Account(
            id = null,
            name = aname,
            level = level,
            isExpanded = false,
            isVisible = true,
            amounts = aggregateBalances(balances)
        )
    }

    private fun aggregateBalances(balances: List<SimpleBalance>): List<AccountAmount> = balances
        .groupBy { it.commodity }
        .map { (commodity, items) ->
            AccountAmount(
                currency = commodity,
                amount = items.sumOf { it.amount.toDouble() }.toFloat()
            )
        }

    data class SimpleBalance(var commodity: String, var amount: Float, var amountStyle: AmountStyle? = null) {
        constructor(commodity: String, amount: Float) : this(commodity, amount, null)
    }
}
